README
NAME
    Regexp::Grammars::Common::String - Some basic String parsing Rules for
    Regexp::Grammars

VERSION
    version 1.000002

SYNOPSIS
        use Regex::Grammars;
        use Regexp::Grammars::Common::String;

        my $re = qr{

            (.*<[String]>)+

            <extends: Regexp::Grammars::Common::String>

        }x; #  don't forget the x!

            ...

        if( $content =~ $re ){
            print Dumper( \%/ ); # Praying Mantis operator!?
        }

DESCRIPTION
    Regexp::Grammars is just too useful to not use, but too pesky and
    confusing for new people.

    Some of the more complex things involve string extraction and
    escape-handling, and I seriously spent the better part 2 hours learning
    how to make this work. So, even if this module is not immediately
    useful, it may serve as an educational tool for others.

    I probably should have delved deeper into the Regexp::Common Family, but
    I couldn't find one in there that did exactly what I wanted.

    At present, this module only provides one rule, "String", but I will
    probably add a few more later.

GRAMMARS
  Regexp::Grammars::Common::String
        <extends: Regexp::Grammars::Common::String>

RULES
  String
    For parsing strings like so:

        "Hello"     => 'Hello'
        "Hel\lo"    => 'Hello'
        "Hel\\lo"   => 'Hel\lo'
        "Hel\"lo"   => 'Hel"lo'

    It should do a reasonable job of picking up strings from files and
    properly returning their parsed contents.

    It made sense to me to drop the excess "\"'s that are used for escaping,
    in order to get a copy of the string as it would be seen to anything
    else that parsed it properly and evaluated the escapes into characters.

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Regexp-Grammars-Common-String

1.000003 2015-12-15T10:59:15Z cb6cbcd

1.000002 2015-12-15T10:56:25Z fdd40fd
 [Bugfix]
 - Avoid complications with Regexp::Grammars 1.043 failing to create match globals ( See RT#110356 ) by explicitly
   declaring them.
 - Closes #1

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑3 -3 (recommends: +1, suggests: ↑2)
 - test: +1 (recommends: ↑1 ↓1)

1.000001 2014-08-15T23:28:10Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - whitespace greatly reduced.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

1.000000 2014-08-07T14:27:54Z
 [00 Major]
 - Significant packaging changes that may affect hand-operated downstream.

 [01 Version Scheme]
 - Versions normalised to x.y{6} instead of x.y{8}
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129
 - This is the primary reason for the 1.x on the box

 [02 Toolkit]
 - Migrate to EUMM
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f
 - "latest" Toolkit dependencies softened to recommendations.

 [Dependencies::Stats]
 - Dependencies changed since 0.01000002, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +48 -1 (recommends: -1, suggests: +1 ↑1)
 - runtime: +1
 - test: +2 -2 (recommends: +3)

 [Minor]
 - Moved to kentnl.org

0.01000002 2011-10-22T13:35:39Z
 - Maintenance Release

 [Dependencies::Stats]
 - Dependencies changed since 0.01000001, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)
 - runtime: +2
 - test: -1

 [Packaging]
 - ship .perltidyrc
 - use Git versions
 - move to @Author::KENTNL
 - Migrate Changelog format to CPAN::Changes
 - Update LICENSE ( Year, Indentation, Address )
 - Move author/release tests to xt/
 - add x_authority
 - point Git repo urls to https:// github

 [Tests]
 - 00-compile.t regenerated.

0.01000001 2010-12-25T01:20:07Z
 - First version. Merry Christmas.


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4206"
  },
  "configure_requires" => {
    "Module::Build" => "0.4206"
  },
  "dist_abstract" => "Some basic String parsing Rules for Regexp::Grammars",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Regexp-Grammars-Common-String",
  "dist_version" => "1.000000",
  "license" => "perl",
  "module_name" => "Regexp::Grammars::Common::String",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Regexp::Grammars" => 0,
    "perl" => "5.010000",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.001003"
  }
);


my %fallback_build_requires = (
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4206",
  "Test::More" => "1.001003"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Regexp/Grammars/Common/String.pm
use 5.010000;
use strict;
use warnings;
use utf8;

package Regexp::Grammars::Common::String;

our $VERSION = '1.000003';

# ABSTRACT: Some basic String parsing Rules for Regexp::Grammars

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Regexp::Grammars;

# Needs to be declared and R:G:1.043 forgets to
# Unrelated to English.pm $MATCH
our $MATCH;    ## no critic (Variables::ProhibitMatchVars)

## no critic (RegularExpressions Documentation Variables::ProhibitUnusedVarsStricter)
my $grammar = qr{
    <grammar: Regexp::Grammars::Common::String>

    <token: String>
        "
        <[MATCH=([^"\\]*)]>+
        (
            \\<[MATCH=(.)]>
            <[MATCH=([^"\\]*)]>
        )*
        "
        <MATCH=(?{ join q{}, @{ $MATCH }})>

}x;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Regexp::Grammars::Common::String - Some basic String parsing Rules for Regexp::Grammars

=head1 VERSION

version 1.000003

=head1 SYNOPSIS

    use Regex::Grammars;
    use Regexp::Grammars::Common::String;

    my $re = qr{

        (.*<[String]>)+

        <extends: Regexp::Grammars::Common::String>

    }x; #  don't forget the x!

        ...

    if( $content =~ $re ){
        print Dumper( \%/ ); # Praying Mantis operator!?
    }

=head1 DESCRIPTION

L<Regexp::Grammars|Regexp::Grammars> is just too useful to not use, but too pesky and confusing for new people.

Some of the more complex things involve string extraction and escape-handling, and I seriously spent the better part 2 hours
learning how to make this work. So, even if this module is not immediately useful, it may serve as an educational tool for
others.

I probably should have delved deeper into the L<Regexp::Common|Regexp::Common> Family, but I couldn't find one in there that
did exactly what I wanted.

At present, this module only provides one rule, L</String>, but I will probably add a few more later.

=head1 GRAMMARS

=head2 Regexp::Grammars::Common::String

    <extends: Regexp::Grammars::Common::String>

=head1 RULES

=head2 String

For parsing strings like so:

    "Hello"     => 'Hello'
    "Hel\lo"    => 'Hello'
    "Hel\\lo"   => 'Hel\lo'
    "Hel\"lo"   => 'Hel"lo'

It should do a reasonable job of picking up strings from files and properly returning their parsed contents.

It made sense to me to drop the excess C<\>'s that are used for escaping, in order to get a copy of the string as
it would be seen to anything else that parsed it properly and evaluated the escapes into characters.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.041.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Some basic String parsing Rules for Regexp::Grammars",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Regexp-Grammars-Common-String",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Regexp::Grammars::Common::String",
  "PREREQ_PM" => {
    "Regexp::Grammars" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "1.000003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Regexp::Grammars" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
